<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Iteratively reweighted least squares (IRLS) procedure in CS-CORE (Rcpp) — CSCORE_IRLS_cpp • CSCORE</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Iteratively reweighted least squares (IRLS) procedure in CS-CORE (Rcpp) — CSCORE_IRLS_cpp"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CSCORE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/CSCORE.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/CSCORE.html">CS-CORE</a></li>
    <li><a class="dropdown-item" href="../articles/covariate_adjustment.html">Covariates adjustment</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"></ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Iteratively reweighted least squares (IRLS) procedure in CS-CORE (Rcpp)</h1>

      <div class="d-none name"><code>CSCORE_IRLS_cpp.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CSCORE_IRLS_cpp</span><span class="op">(</span></span>
<span>  <span class="va">X</span>,</span>
<span>  <span class="va">seq_depth</span>,</span>
<span>  covariates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  post_process <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  covariate_level <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>  adjust_setting <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="cn">T</span>, var <span class="op">=</span> <span class="cn">T</span>, covar <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>  IRLS_par <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>n_iter <span class="op">=</span> <span class="fl">10</span>, eps <span class="op">=</span> <span class="fl">0.05</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, conv <span class="op">=</span> <span class="st">"q95"</span><span class="op">)</span>,</span>
<span>  return_all <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="source">Source<a class="anchor" aria-label="anchor" href="#source"></a></h2>
    <p>Su, C., Xu, Z., Shan, X., Cai, B., Zhao, H., &amp; Zhang, J. (2023).
Cell-type-specific co-expression inference from single cell RNA-sequencing data.
<em>Nature Communications</em>.
doi: <a href="https://doi.org/10.1038/s41467-023-40503-7" class="external-link">https://doi.org/10.1038/s41467-023-40503-7</a></p>
    </div>
    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">X<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A numeric matrix of UMI counts (<code>n x p</code>),
where <code>n</code> is the number of cells and <code>p</code> is the number of genes.</p></dd>


<dt id="arg-seq-depth">seq_depth<a class="anchor" aria-label="anchor" href="#arg-seq-depth"></a></dt>
<dd><p>A numeric vector of sequencing depths of length <code>n</code>.</p></dd>


<dt id="arg-covariates">covariates<a class="anchor" aria-label="anchor" href="#arg-covariates"></a></dt>
<dd><p>Optional. A numeric matrix of covariates (<code>n x K</code>) to be adjusted for in the moment-based regressions.
Can be a length n vector if \(K=1\).
If <code>NULL</code>, no covariates are adjusted for in the regression.
Defaults to <code>NULL</code>.</p></dd>


<dt id="arg-post-process">post_process<a class="anchor" aria-label="anchor" href="#arg-post-process"></a></dt>
<dd><p>Optional. Logical; whether to rescale the estimated co-expressions to lie between –1 and 1.
Defaults to <code>TRUE</code>.</p></dd>


<dt id="arg-covariate-level">covariate_level<a class="anchor" aria-label="anchor" href="#arg-covariate-level"></a></dt>
<dd><p>Optional. A character string indicating whether covariates are assumed to affect
the underlying gene expression levels (<code>"z"</code>) or the observed counts (<code>"x"</code>).
See the <em>Details</em> section for further explanation.
Defaults to <code>"z"</code>.</p></dd>


<dt id="arg-adjust-setting">adjust_setting<a class="anchor" aria-label="anchor" href="#arg-adjust-setting"></a></dt>
<dd><p>Optional. A named logical vector of length 3;
whether to adjust for covariates at the mean, variance, and covariance level.
Must be named <code>c("mean", "var", "covar")</code>.
Defaults to <code>c(mean = T,var = T, covar = T)</code>.</p></dd>


<dt id="arg-irls-par">IRLS_par<a class="anchor" aria-label="anchor" href="#arg-irls-par"></a></dt>
<dd><p>Optional. A named list of length 4 specifying parameters for the IRLS algorithm:</p><dl><dt><code>n_iter</code></dt>
<dd><p>Maximum number of iterations.</p></dd>

<dt><code>eps</code></dt>
<dd><p>Convergence threshold for log-ratio change <code>delta</code>, computed as <code>abs(log(theta / theta_prev))</code>.</p></dd>

<dt><code>verbose</code></dt>
<dd><p>Logical; whether to print the convergence metric (<code>delta</code>) at each iteration.</p></dd>

<dt><code>conv</code></dt>
<dd><p>Character string; determine convergence based on <code>q95</code> (95\<!-- % quantile) or \code{max} of \code{delta}.} --></p></dd>

Defaults to list(n_iter = 10, eps = 0.05, verbose = FALSE, conv = "q95").
</dl><p>return_allLogical; whether to return all estimates, including the effect sizes for covariates.
Defaults to <code>FALSE</code>.</p></dd>

</dl><p>A list containing the following components:
</p><dl><dt>est</dt>
<dd><p>A \(p \times p\) matrix of co-expression estimates.</p></dd><dt>p_value</dt>
<dd><p>A \(p \times p\) matrix of p-values.</p></dd><dt>test_stat</dt>
<dd><p>A \(p \times p\) matrix of test statistics evaluating the statistical significance of co-expression.</p></dd><dt>mu_beta</dt>
<dd><p>A \(k \times p\) matrix of regression coefficients from the mean model. Returned if <code>return_all = TRUE</code>.</p></dd><dt>sigma2_beta</dt>
<dd><p>A \(k \times p\) matrix of regression coefficients from the variance model. Returned if <code>return_all = TRUE</code>.</p></dd><dt>cov_beta</dt>
<dd><p>A \(k \times p \times p\) array of regression coefficients from the covariance model. Returned if <code>return_all = TRUE</code>.</p></dd></dl>


Estimate and test cell-type-specific co-expression using the IRLS procedure with optional covariate adjustment.


Let \(x_{ij}\) denote the UMI count of gene \(j\) in cell \(i\);
\(s_i\) denote the sequencing depth;
\(\mu_j,\sigma_{jj}, \sigma_{jj'}\) denote the mean, variance and covariance;
\(c_{ik}\) denote additional covariate \(k\) for cell \(i\) (e.g. disease status or cellular state).
The procedure consists of two main steps:
<ol><li><p><strong>Mean and variance estimation:</strong> Estimate gene-specific mean and variance parameters
using two moment-based regressions:</p><ul><li><p><strong>Mean model:</strong> \(x_{ij} = s_i (\mu_j + \sum_k c_{ik} \beta_k) + \epsilon_{ij}\)</p></li>
<li><p><strong>Variance model:</strong> \((x_{ij} - s_i \mu_{ij})^2 = s_i \mu_{ij} + s_i^2 (\sigma_{jj} + \sum_k c_{ik} \gamma_k) + \eta_{ij}\),
where \(\mu_{ij} = \mu_j + \sum_k c_{ik} \beta_k\)</p></li>
</ul></li>
<li><p><strong>Covariance estimation and hypothesis testing:</strong> Estimate gene-gene covariance
and compute test statistics to assess the statistical significance of gene co-expression using a third moment-based regression:</p><ul><li><p><strong>Covariance model:</strong> \((x_{ij} - s_i \mu_{ij})(x_{ij'} - s_i \mu_{ij'}) = s_i^2 (\sigma_{jj'} + \sum_k c_{ik} \theta_k) + \xi_{ijj'}\)</p></li>
</ul></li>
</ol>
We note that
<ol><li><p>The formulation above assumes that the covariates alter the mean / variance / covariance of underlying gene expression,
rather than observed counts. If you believe that the covariates directly affect the observed counts independent of underlying gene expression
(e.g. \(x_{ij}|z_{ij} \sim \text{Poisson}(s_i z_{ij}+\sum_k c_{ik} \beta_k)\) or \(x_{ij} = s_i \mu_j  + \sum_k c_{ik} \beta_k + \epsilon_{ij}\)),
please specify <code>covariate_level="x"</code>.</p></li>
<li><p>The original CS-CORE published in <a href="https://doi.org/10.1038/s41467-023-40503-7" class="external-link">https://doi.org/10.1038/s41467-023-40503-7</a>
did not consider adjusting for covariates \(c_{ik}\)'s. This is equivalent to setting <code>covariates</code> to <code>NULL</code>.</p></li>
</ol>
Note: This is an R wrapper for the <code>CSCORE_IRLS_cpp_impl()</code> function implemented in Rcpp.


## Toy example:
## run CSCORE on a simulated independent gene pair
cscore_example &lt;- CSCORE_IRLS_cpp(ind_gene_pair$counts, ind_gene_pair$seq_depths)## Estimated co-expression between two genes
cscore_example$est[1,2]
# close to 0: 0.007820124## p-values
cscore_example$p_value[1,2]
# not significant: 0.961981
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su, Zichun Xu, Xinning Shan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

