<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>CS-CORE for cell-type-specific co-expression analysis • CSCORE</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CS-CORE for cell-type-specific co-expression analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">CSCORE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/CSCORE.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/CSCORE.html">CS-CORE</a></li>
    <li><a class="dropdown-item" href="../articles/covariate_adjustment.html">Covariates adjustment</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav"></ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>CS-CORE for cell-type-specific co-expression analysis</h1>
            
      

      <div class="d-none name"><code>CSCORE.Rmd</code></div>
    </div>

    
    
<p>This vignette shows an example of applying <code>CS-CORE</code> to
infer cell-type-specific co-expression networks and extracting
co-expressed gene modules that are enriched for biological functions in
cell types.</p>
<div class="section level2">
<h2 id="load-packages-and-data">1. Load packages and data<a class="anchor" aria-label="anchor" href="#load-packages-and-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://changsubiostats.github.io/CS-CORE/">CSCORE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<p>In this vignette, we use the single cell RNA-sequencing data on
Peripheral blood mononuclear cells (PBMC) from COVID patients and
healthy controls from <a href="https://www.nature.com/articles/s41591-020-0944-y" class="external-link">Wilk et
al.</a>, which were also studied in our <a href="https://www.nature.com/articles/s41467-023-40503-7" class="external-link">paper</a>.
This data set can be downloaded via the following bash script</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">wget</span> https://hosted-matrices-prod.s3-us-west-2.amazonaws.com/Single_cell_atlas_of_peripheral_immune_response_to_SARS_CoV_2_infection-25/blish_covid.seu.rds</span></code></pre></div>
<p>After downloading blish_covid.seu.rds, we load it into the R
session</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'blish_covid.seu.rds'</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="co"># update the obsolete Seurat object</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="select-cell-types-and-gene-sets-to-study">2. Select cell types and gene sets to study<a class="anchor" aria-label="anchor" href="#select-cell-types-and-gene-sets-to-study"></a>
</h2>
<p>In this example, we focus on B cells and infer the B cell–specific
co-expression network. In general, we strongly recommend subsetting to
the cell type of interest before inferring co-expression. Otherwise, the
estimates may be affected by <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC4579334/#:~:text=(A)%20Simpson's%20Paradox%20describes%20" class="external-link">Simpson’s
paradox</a>, capturing associations driven by differences between cell
types (e.g.correlations between cell type marker genes), rather than
gene-gene co-expression within a homogeneous cell population.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_B</span> <span class="op">=</span> <span class="va">pbmc</span><span class="op">[</span>,<span class="va">pbmc</span><span class="op">$</span><span class="va">cell.type.coarse</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">'B'</span><span class="op">]</span></span></code></pre></div>
<p>Depending on the biological question of interest, one may choose to
study the co-expression network for any gene set. Here, we chose to
infer the co-expression network for the genes with meaningful expression
levels in B cells (top 5000 among 26361 genes). There are several
reasons for our choice:</p>
<ol style="list-style-type: decimal">
<li><p>All genes with moderate to high expression levels provides a
comprehensive and unbiased set of genes that could have meaningful
biological functions in a cell type.</p></li>
<li><p>If the genes have much lower expression levels, it would be
statistically more challenging and biologically less interesting to
infer their co-expressions, as these genes might have almost all UMI
counts equal to 0.</p></li>
</ol>
<p>In general, it will be up to the users’s choice to select the gene
sets to study. We recommend choosing the gene sets that are of interest
to your application.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean_exp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">pbmc_B</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">/</span><span class="va">pbmc_B</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span></span>
<span><span class="va">genes_selected</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort.int</a></span><span class="op">(</span><span class="va">mean_exp</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5000</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-cs-core-to-infer-cell-type-specific-co-expression-network-on-the-specified-gene-set">3. Run <code>CS-CORE</code> to infer cell-type-specific
co-expression network on the specified gene set<a class="anchor" aria-label="anchor" href="#run-cs-core-to-infer-cell-type-specific-co-expression-network-on-the-specified-gene-set"></a>
</h2>
<p>We further subset the B cells to those from healthy control subjects
in order to study B-cell specific co-expression network among healthy
control B cells.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_B_healthy</span> <span class="op">&lt;-</span> <span class="va">pbmc_B</span><span class="op">[</span>, <span class="va">pbmc_B</span><span class="op">$</span><span class="va">Status</span> <span class="op">==</span> <span class="st">"Healthy"</span><span class="op">]</span></span></code></pre></div>
<p>Run <code>CS-CORE</code> with the subsetted Seurat object and a gene
set of interest. We note that CSCORE operates on the raw UMI counts,
i.e. <code>object[['RNA']]@counts</code> for the Seurat object
<code>object</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CSCORE_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CSCORE.html">CSCORE</a></span><span class="op">(</span><span class="va">pbmc_B_healthy</span>, genes <span class="op">=</span> <span class="va">genes_selected</span><span class="op">)</span></span>
<span><span class="co"># faster yet more memory intensive:</span></span>
<span><span class="co"># CSCORE_result &lt;- CSCORE(pbmc_B_healthy, genes = genes_selected, IRLS_version = 'bash_R')</span></span></code></pre></div>
<p>Since version 1.0.1 (updated June 2025), CSCORE provides a new Rcpp
implementation of moment-based regressions that offers improved memory
efficiency and supports covariate adjustment (see <a href="covariate_adjustment.html">Covariate adjustment</a>). The original
implementation at the time of publication (version 0.0.0.9) was written
in base R, and while it is approximately 10 times faster, it is also
significantly more memory intensive (by a factor of 10–100). We
recommend choosing between the two implementations based on available
memory resources and the need for computational speed and covariate
adjustment.</p>
</div>
<div class="section level2">
<h2 id="downstream-analysis-on-the-co-expression-network">4. Downstream analysis on the co-expression network<a class="anchor" aria-label="anchor" href="#downstream-analysis-on-the-co-expression-network"></a>
</h2>
<div class="section level3">
<h3 id="extract-co-expressed-gene-module">4.1 Extract co-expressed gene module<a class="anchor" aria-label="anchor" href="#extract-co-expressed-gene-module"></a>
</h3>
<p>Given the <code>CS-CORE</code> <span class="math inline">\(p\)</span>-values, we first set co-expressions
that are not statistically significant (Benjamini &amp;
Hochberg-adjusted <span class="math inline">\(p\)</span>-values <span class="math inline">\(&gt;0.05\)</span>) to 0.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Obtain CS-CORE co-expression estimates</span></span>
<span><span class="va">CSCORE_coexp</span> <span class="op">&lt;-</span> <span class="va">CSCORE_result</span><span class="op">$</span><span class="va">est</span></span>
<span></span>
<span><span class="co"># Obtain BH-adjusted p values</span></span>
<span><span class="va">CSCORE_p</span> <span class="op">&lt;-</span> <span class="va">CSCORE_result</span><span class="op">$</span><span class="va">p_value</span></span>
<span><span class="va">p_matrix_BH</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_selected</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_selected</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_matrix_BH</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">upper.tri</a></span><span class="op">(</span><span class="va">p_matrix_BH</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">CSCORE_p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">upper.tri</a></span><span class="op">(</span><span class="va">CSCORE_p</span><span class="op">)</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="va">p_matrix_BH</span> <span class="op">&lt;-</span> <span class="va">p_matrix_BH</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">p_matrix_BH</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set co-expression entires with BH-adjusted p-values greater than 0.05 to 0</span></span>
<span><span class="va">CSCORE_coexp</span><span class="op">[</span><span class="va">p_matrix_BH</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
<p>Next, based on the thresholded co-expression matrix, we apply <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559" class="external-link">WGCNA</a>
to extract co-expressed gene modules. In particular, we use
<code>CS-CORE</code> estimates to measure co-expressions for single cell
RNA-sequencing data, which replace the Pearson correlations used in
traditional WGNCA workflow, that suffer from inflated false positives
and attenuation bias on single cell data as demonstrated in our <a href="https://www.biorxiv.org/content/10.1101/2022.12.13.520181v1" class="external-link">manuscript</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">WGCNA</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"WGCNA"</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">WGCNA</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">WGCNA</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the adjacency matrix based on the co-expression matrix</span></span>
<span><span class="va">adj</span> <span class="op">=</span> <span class="fu">WGCNA</span><span class="fu">::</span><span class="fu">adjacency.fromSimilarity</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">CSCORE_coexp</span><span class="op">)</span>, power <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Compute the topological overlap matrix</span></span>
<span><span class="va">TOM</span> <span class="op">=</span> <span class="fu">WGCNA</span><span class="fu">::</span><span class="fu">TOMsimilarity</span><span class="op">(</span><span class="va">adj</span><span class="op">)</span></span>
<span><span class="va">dissTOM</span> <span class="op">=</span> <span class="fl">1</span><span class="op">-</span><span class="va">TOM</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dissTOM</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dissTOM</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes_selected</span></span>
<span><span class="co"># Run hierarchical clustering as in the WGCNA workflow</span></span>
<span><span class="va">hclust_dist</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="va">dissTOM</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"average"</span><span class="op">)</span> </span>
<span><span class="va">memb</span> <span class="op">=</span> <span class="fu">dynamicTreeCut</span><span class="fu">::</span><span class="fu">cutreeDynamic</span><span class="op">(</span>dendro <span class="op">=</span> <span class="va">hclust_dist</span>, </span>
<span>                     distM <span class="op">=</span> <span class="va">dissTOM</span>, </span>
<span>                     deepSplit <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                     pamRespectsDendro <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     minClusterSize <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co"># For more instructions on how to tune the parameters in the WGCNA workflow,</span></span>
<span><span class="co"># please refer to https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">memb</span><span class="op">)</span> <span class="op">=</span> <span class="va">genes_selected</span></span>
<span><span class="va">memb_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">memb</span><span class="op">)</span></span>
<span><span class="va">module_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">memb</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i_k</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">memb</span> <span class="op">==</span> <span class="va">i_k</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>One can also apply other clustering methods to extract co-expressed
gene modules.</p>
</div>
<div class="section level3">
<h3 id="functional-enrichment-analysis">4.2 Functional enrichment analysis<a class="anchor" aria-label="anchor" href="#functional-enrichment-analysis"></a>
</h3>
<p>Gene Ontology (GO) enrichment analysis is one of the common
downstream functional enrichment analyses for interpreting the
biological pathways implied by co-expressed gene modules. Here, we
showcase the GO pathways enriched in <code>CS-CORE</code> co-expressed
gene modules using the R implementation from <a href="https://pubmed.ncbi.nlm.nih.gov/34557778/" class="external-link">Wu et al.</a>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set all genes in clustering analysis as background,</span></span>
<span><span class="co"># such that the enrichment result of any module is not attributed to its high expression levels.</span></span>
<span><span class="va">universe</span> <span class="op">&lt;-</span> <span class="va">genes_selected</span></span>
<span></span>
<span><span class="co"># Filter GO terms based on BH-adjusted p values &lt; 0.05</span></span>
<span><span class="co">####</span></span>
<span><span class="co">## Note: the following codes can take a long time to run as </span></span>
<span><span class="co">## in this example there are more than 100 co-expressed gene modules from WGCNA</span></span>
<span><span class="co">####</span></span>
<span><span class="va">ego_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">module_list</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">module_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,</span>
<span>         OrgDb <span class="op">=</span> <span class="st">'org.Hs.eg.db'</span>, <span class="co"># human</span></span>
<span>         keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>         ont <span class="op">=</span> <span class="st">"ALL"</span>,</span>
<span>         pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>         universe <span class="op">=</span> <span class="va">universe</span>,</span>
<span>         pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>There are in total 144 gene modules inferred by WGCNA. For
illustrative purposes, we focus on the modules with the strongest
enrichment signals (with at least one GO term having adjusted <span class="math inline">\(p\)</span>-value smaller than <span class="math inline">\(10^{-3}\)</span> and with at least 10 enriched GO
terms) and print the top 3 GO terms.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_enrich_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ego_result</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> </span>
<span>  <span class="op">(</span><span class="va">x</span><span class="op">@</span><span class="va">result</span><span class="op">$</span><span class="va">p.adjust</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.001</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">&gt;</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">top_enrich_go</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">top_enrich_clusters</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">ego_result</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">result</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">top_enrich_go</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">top_enrich_go</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Description'</span>, <span class="st">'GeneRatio'</span>, <span class="st">'p.adjust'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'\n'</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt;                             Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0002181      cytoplasmic translation    83/232 1.289669e-80</span></span>
<span><span class="co">#&gt; GO:0006412                  translation    94/232 3.123690e-43</span></span>
<span><span class="co">#&gt; GO:0043043 peptide biosynthetic process    94/232 3.123690e-43</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                                Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0030029    actin filament-based process    21/107 0.0002847674</span></span>
<span><span class="co">#&gt; GO:0030036 actin cytoskeleton organization    20/107 0.0002847674</span></span>
<span><span class="co">#&gt; GO:0007015     actin filament organization    16/107 0.0004358998</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                                                                                  Description</span></span>
<span><span class="co">#&gt; GO:0019886 antigen processing and presentation of exogenous peptide antigen via MHC class II</span></span>
<span><span class="co">#&gt; GO:0002495           antigen processing and presentation of peptide antigen via MHC class II</span></span>
<span><span class="co">#&gt; GO:0002399                                             MHC class II protein complex assembly</span></span>
<span><span class="co">#&gt;            GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0019886    14/109 7.766970e-16</span></span>
<span><span class="co">#&gt; GO:0002495    14/109 1.036171e-15</span></span>
<span><span class="co">#&gt; GO:0002399    12/109 1.036171e-15</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                          Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0002250  adaptive immune response     13/37 4.972114e-05</span></span>
<span><span class="co">#&gt; GO:0006910 phagocytosis, recognition      6/37 4.114611e-04</span></span>
<span><span class="co">#&gt; GO:0042113         B cell activation     10/37 4.114611e-04</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                             Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0035455 response to interferon-alpha      4/25 0.0001664956</span></span>
<span><span class="co">#&gt; GO:0009615            response to virus      8/25 0.0004724418</span></span>
<span><span class="co">#&gt; GO:0051607    defense response to virus      7/25 0.0004724418</span></span></code></pre></div>
<p>At this point, we have reproduced the results in our <a href="https://doi.org/10.1038/s41467-023-40503-7" class="external-link">manuscript</a>, Table
S6.</p>
<p>This concludes our vignette of using <code>CS-CORE</code> to infer
cell-type-specific co-expression networks and a typical pipeline for
extracting co-expressed gene modules and performing functional
enrichment analysis.</p>
<p>One can also perform a differential co-expression analysis based on
the codes provided here. For example, the inferred network of healthy B
cells can be constrasted to the network inferred with B cells from the
COVID-19 patients to study dysregulation in B cells’ co-expression due
to COVID-19 infection. For more details please refer the the methods in
our <a href="https://doi.org/10.1038/s41467-023-40503-7" class="external-link">manuscript</a>.</p>
<p><strong>Stay tuned!</strong> We are also working on a pipeline for
cell-type-specific module-trait association analyses with single cell
RNA-seq data based on <code>CS-CORE</code> developed here.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chang Su, Zichun Xu, Xinning Shan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
