<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="CSCORE">
<title>CS-CORE for cell-type-specific co-expression analysis • CSCORE</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CS-CORE for cell-type-specific co-expression analysis">
<meta property="og:description" content="CSCORE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CSCORE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/CSCORE.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>CS-CORE for cell-type-specific co-expression analysis</h1>
            
      
      
      <div class="d-none name"><code>CSCORE.Rmd</code></div>
    </div>

    
    
<p>This vignette shows an example of applying <code>CS-CORE</code> to
infer cell-type-specific co-expression networks and extracting
co-expressed gene modules for functional enrichment analyses.</p>
<div class="section level2">
<h2 id="load-packages-and-data">1. Load packages and data<a class="anchor" aria-label="anchor" href="#load-packages-and-data"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://changsubiostats.github.io/CS-CORE/">CSCORE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<p>In this vignette, we use the single cell RNA-sequencing data on
Peripheral blood mononuclear cells (PBMC) from COVID patients and
healthy controls from <a href="https://www.nature.com/articles/s41591-020-0944-y" class="external-link">Wilk et
al.</a>, which was also studied in our <a href="https://www.biorxiv.org/content/10.1101/2022.12.13.520181v1" class="external-link">manuscript</a>.
This data set can be downloaded via</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://hosted-matrices-prod.s3-us-west-2.amazonaws.com/Single_cell_atlas_of_peripheral_immune_response_to_SARS_CoV_2_infection-25/blish_covid.seu.rds</span></code></pre></div>
<p>After downloading blish_covid.seu.rds, we load it into the R
session</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'data/blish_covid.seu.rds'</span><span class="op">)</span></span>
<span><span class="co"># Use the original UMI counts stored in Assay 'RNA'</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'RNA'</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="select-cell-types-and-gene-sets-to-study">2. Select cell types and gene sets to study<a class="anchor" aria-label="anchor" href="#select-cell-types-and-gene-sets-to-study"></a>
</h2>
<p>In this example, we focus on B cells and infer the B cell-specific
co-expression network.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_B</span> <span class="op">=</span> <span class="va">pbmc</span><span class="op">[</span>,<span class="va">pbmc</span><span class="op">$</span><span class="va">cell.type.coarse</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">'B'</span><span class="op">]</span></span></code></pre></div>
<p>Depending on the biological question of interest, one may choose to
study the co-expression network for any particular gene set. Here, we
chose to infer the co-expression network for the genes with meaningful
expression levels in B cells (top 5000 among 26361 genes). There are
several reasons for our choice:</p>
<ol style="list-style-type: decimal">
<li><p>All genes with moderate to high expression levels provides a
comprehensive and unbiased set of genes that could have meaningful
biological functions in a cell type.</p></li>
<li><p>It is statistically challenging to infer co-expression for genes
that have almost all UMI counts equal to 0.</p></li>
<li><p>It is biologically less interesting to study the co-expressions
if the genes are not expressed in a cell type.</p></li>
</ol>
<p>However, it will be up to the users’s choice to select the gene sets
to study. We recommend choosing the gene sets that are of interest to
your application.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mean_exp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">pbmc_B</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">/</span><span class="va">pbmc_B</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span></span>
<span><span class="va">genes_selected</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort.int</a></span><span class="op">(</span><span class="va">mean_exp</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5000</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-cs-core-to-infer-cell-type-specific-co-expression-network-on-the-specified-gene-set">3. Run <code>CS-CORE</code> to infer cell-type-specific
co-expression network on the specified gene set<a class="anchor" aria-label="anchor" href="#run-cs-core-to-infer-cell-type-specific-co-expression-network-on-the-specified-gene-set"></a>
</h2>
<p>We further subset the B cells to those from healthy control subjects
in order to study B-cell specific co-expression network from healthy
control cells. Though not demonstrated in this vignette, the inferred
network can further be constrasted to networks inferred with B cells
from the COVID-19 patients to study dysregulation in B cells’
co-expression due to COVID-19 infection.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_B_healthy</span> <span class="op">&lt;-</span> <span class="va">pbmc_B</span><span class="op">[</span>, <span class="va">pbmc_B</span><span class="op">$</span><span class="va">Status</span> <span class="op">==</span> <span class="st">"Healthy"</span><span class="op">]</span></span></code></pre></div>
<p>Run <code>CS-CORE</code> with the subsetted Seurat object and a gene
set of interest.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CSCORE_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CSCORE.html">CSCORE</a></span><span class="op">(</span><span class="va">pbmc_B_healthy</span>, genes <span class="op">=</span> <span class="va">genes_selected</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downstream-analysis-on-the-co-expression-network">4. Downstream analysis on the co-expression network<a class="anchor" aria-label="anchor" href="#downstream-analysis-on-the-co-expression-network"></a>
</h2>
<div class="section level3">
<h3 id="extract-co-expressed-gene-module">4.1 Extract co-expressed gene module<a class="anchor" aria-label="anchor" href="#extract-co-expressed-gene-module"></a>
</h3>
<p>Given the <code>CS-CORE</code> <span class="math inline">\(p\)</span>-values, we first set co-expressions
that are not statistically significant (Benjamini &amp;
Hochberg-adjusted <span class="math inline">\(p\)</span>-values <span class="math inline">\(&gt;0.05\)</span>) to 0.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Obtain CS-CORE co-expression estimates</span></span>
<span><span class="va">CSCORE_coexp</span> <span class="op">&lt;-</span> <span class="va">CSCORE_result</span><span class="op">$</span><span class="va">est</span></span>
<span></span>
<span><span class="co"># Obtain BH-adjusted p values</span></span>
<span><span class="va">CSCORE_p</span> <span class="op">&lt;-</span> <span class="va">CSCORE_result</span><span class="op">$</span><span class="va">p_value</span></span>
<span><span class="va">p_matrix_BH</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_selected</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_selected</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_matrix_BH</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">upper.tri</a></span><span class="op">(</span><span class="va">p_matrix_BH</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">CSCORE_p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/lower.tri.html" class="external-link">upper.tri</a></span><span class="op">(</span><span class="va">CSCORE_p</span><span class="op">)</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span><span class="va">p_matrix_BH</span> <span class="op">&lt;-</span> <span class="va">p_matrix_BH</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">p_matrix_BH</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set co-expression entires with BH-adjusted p-values greater than 0.05 to 0</span></span>
<span><span class="va">CSCORE_coexp</span><span class="op">[</span><span class="va">p_matrix_BH</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
<p>Next, based on the thresholded co-expression matrix, we apply <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-559" class="external-link">WGCNA</a>
to extract co-expressed gene modules. In particular, we use
<code>CS-CORE</code> estimates to measure co-expressions for single cell
RNA-sequencing data, while the traditional WGNCA workflow was designed
for bulk expression data and used Pearson correlations to measure
co-expressions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="http://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/" class="external-link">WGCNA</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"WGCNA"</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/" class="external-link">WGCNA</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/" class="external-link">WGCNA</a></span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the adjacency matrix based on the co-expression matrix</span></span>
<span><span class="va">adj</span> <span class="op">=</span> <span class="fu">WGCNA</span><span class="fu">::</span><span class="fu">adjacency.fromSimilarity</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">CSCORE_coexp</span><span class="op">)</span>, power <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># Compute the topological overlap matrix</span></span>
<span><span class="va">TOM</span> <span class="op">=</span> <span class="fu">WGCNA</span><span class="fu">::</span><span class="fu">TOMsimilarity</span><span class="op">(</span><span class="va">adj</span><span class="op">)</span></span>
<span><span class="va">dissTOM</span> <span class="op">=</span> <span class="fl">1</span><span class="op">-</span><span class="va">TOM</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dissTOM</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dissTOM</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes_selected</span></span>
<span><span class="co"># Run hierarchical clustering as in the WGCNA workflow</span></span>
<span><span class="va">hclust_dist</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="va">dissTOM</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"average"</span><span class="op">)</span> </span>
<span><span class="va">memb</span> <span class="op">=</span> <span class="fu">dynamicTreeCut</span><span class="fu">::</span><span class="fu">cutreeDynamic</span><span class="op">(</span>dendro <span class="op">=</span> <span class="va">hclust_dist</span>, </span>
<span>                     distM <span class="op">=</span> <span class="va">dissTOM</span>, </span>
<span>                     deepSplit <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                     pamRespectsDendro <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                     minClusterSize <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co"># For more instructions on how to tune the parameters in the WGCNA workflow,</span></span>
<span><span class="co"># please refer to https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">memb</span><span class="op">)</span> <span class="op">=</span> <span class="va">genes_selected</span></span>
<span><span class="va">memb_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">memb</span><span class="op">)</span></span>
<span><span class="va">module_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">memb</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i_k</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">memb</span> <span class="op">==</span> <span class="va">i_k</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>One can also apply other clustering methods to extract co-expressed
gene modules.</p>
</div>
<div class="section level3">
<h3 id="functional-enrichment-analysis">4.2 Functional enrichment analysis<a class="anchor" aria-label="anchor" href="#functional-enrichment-analysis"></a>
</h3>
<p>Gene Ontology (GO) enrichment analysis is one of the common
downstream functional enrichment analyses for interpreting the
biological pathways implied by co-expressed gene modules. Here, we
showcase the GO pathways implied by <code>CS-CORE</code> co-expressed
gene modules using the R implementation from <a href="https://pubmed.ncbi.nlm.nih.gov/34557778/" class="external-link">Wu et al.</a>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set all genes in clustering analysis as background,</span></span>
<span><span class="co"># such that the enrichment result of any module is not attributed to its high expression levels.</span></span>
<span><span class="va">universe</span> <span class="op">&lt;-</span> <span class="va">genes_selected</span></span>
<span></span>
<span><span class="co"># Filter GO terms based on BH-adjusted p values &lt; 0.05</span></span>
<span><span class="co">####</span></span>
<span><span class="co">## Note: the following code can take a long time to run as </span></span>
<span><span class="co">## in this example there are 144 co-expressed gene modules from WGCNA</span></span>
<span><span class="co">####</span></span>
<span><span class="va">ego_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">module_list</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">module_list</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>,</span>
<span>         OrgDb <span class="op">=</span> <span class="st">'org.Hs.eg.db'</span>, <span class="co"># human</span></span>
<span>         keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>         ont <span class="op">=</span> <span class="st">"ALL"</span>,</span>
<span>         pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>         universe <span class="op">=</span> <span class="va">universe</span>,</span>
<span>         pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>There are in total 144 gene modules inferred by WGCNA. For
illustrative purposes, we focus on the modules with the strongest
enrichment signals (with at least one GO term having adjusted <span class="math inline">\(p\)</span>-value smaller than <span class="math inline">\(10^{-3}\)</span>; with at least 10 enriched GO
terms) and print the top 3 GO terms.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_enrich_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ego_result</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> </span>
<span>  <span class="op">(</span><span class="va">x</span><span class="op">@</span><span class="va">result</span><span class="op">$</span><span class="va">p.adjust</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.001</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">&gt;</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">top_enrich_go</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">top_enrich_clusters</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="va">ego_result</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">result</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">top_enrich_go</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">top_enrich_go</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Description'</span>, <span class="st">'GeneRatio'</span>, <span class="st">'p.adjust'</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'\n'</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt;                             Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0002181      cytoplasmic translation    83/232 1.289669e-80</span></span>
<span><span class="co">#&gt; GO:0006412                  translation    94/232 3.123690e-43</span></span>
<span><span class="co">#&gt; GO:0043043 peptide biosynthetic process    94/232 3.123690e-43</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                                Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0030029    actin filament-based process    21/107 0.0002847674</span></span>
<span><span class="co">#&gt; GO:0030036 actin cytoskeleton organization    20/107 0.0002847674</span></span>
<span><span class="co">#&gt; GO:0007015     actin filament organization    16/107 0.0004358998</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                                                                                  Description</span></span>
<span><span class="co">#&gt; GO:0019886 antigen processing and presentation of exogenous peptide antigen via MHC class II</span></span>
<span><span class="co">#&gt; GO:0002495           antigen processing and presentation of peptide antigen via MHC class II</span></span>
<span><span class="co">#&gt; GO:0002399                                             MHC class II protein complex assembly</span></span>
<span><span class="co">#&gt;            GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0019886    14/109 7.766970e-16</span></span>
<span><span class="co">#&gt; GO:0002495    14/109 1.036171e-15</span></span>
<span><span class="co">#&gt; GO:0002399    12/109 1.036171e-15</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                          Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0002250  adaptive immune response     13/37 4.972114e-05</span></span>
<span><span class="co">#&gt; GO:0006910 phagocytosis, recognition      6/37 4.114611e-04</span></span>
<span><span class="co">#&gt; GO:0042113         B cell activation     10/37 4.114611e-04</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                             Description GeneRatio     p.adjust</span></span>
<span><span class="co">#&gt; GO:0035455 response to interferon-alpha      4/25 0.0001664956</span></span>
<span><span class="co">#&gt; GO:0009615            response to virus      8/25 0.0004724418</span></span>
<span><span class="co">#&gt; GO:0051607    defense response to virus      7/25 0.0004724418</span></span></code></pre></div>
<p>At this point, we have reproduced the results in our <a href="https://www.biorxiv.org/content/10.1101/2022.12.13.520181v1" class="external-link">manuscript</a>,
Table S6.</p>
<p>This concludes our vignette of using <code>CS-CORE</code> to infer
cell-type-specific co-expression networks and a typical analysis
pipeline for extracting co-expressed gene modules and performing
functional enrichment analysis.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Chang Su, Zichun Xu, Xinning Shan.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
